(react-slider) - Adding the initial Spec.md for the converged Slider component (#19051)
microsoft/fluentui: packages/react-slider/Spec.md

    | `End` | Sets value to the `max` prop. |
  - _Cursor_
    `mousedown` sets the current value immediately.
    `mousemove` is attached to the window element on `mousedown` and watches for move events. Value is updated accordingly.
    `mouseup` removes the `mousemove` event.
  - _Touch_
    Handles the same events as the _Cursor_
  - _Screen readers_
    - **`root`:**
      - renders `as` div
      - handles native props expected from the element type in `as`
    - **`thumb`:**
      - `tabindex = 0`
      - `role = "slider"`
      - `aria-valuemin = min` and `aria-valuemax = max`
      - `aria-valuenow` represents the current value.

## Accessibility

`Slider` design pattern: [W3 Slider](https://www.w3.org/TR/wai-aria-1.1/#slider)

## Topics for Discussion

### RangedSlider Separated

It may be more readable and potentially better to split the `RangedSlider` and `Slider` components apart from each other as they have very different use cases and accessibility concerns.

### Removing Label

If there is a way to remove the Label while preserving accessibility it could greatly benefit the render structure (Removes container and label).

> ### Adding a HTML Range Slider for Accessibility
> 
> Adding a hidden range slider can potentially assist in handling the Slider's accessibility behavior. It is currently being planned to be added once all API features have been fully implemented.

Reviewer: nit: 
```suggestion
### Adding a HTML input of type range for Accessibility

Adding a hidden input of type range can potentially assist in handling the Slider's accessibility behavior. It is currently being planned to be added once all API features have been fully implemented.
```

### Visual behavior props

| Name              | <img src="https://img.shields.io/badge/Used%20in-v0-orange" alt="drawing" width="200"/> | <img src="https://img.shields.io/badge/Used%20in-v8-blue" alt="drawing" width="200"/> | Description                                                                                                   |
| ----------------- | --------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------- |
| label             | x                                                                                       | &check;                                                                               | The description label of the **Slider**.                                                                      |

> | valueLabel        | x                                                                                       | &check;                                                                               | The current value or unique format to be shown for the **Slider's** `value label`.                            |

Reviewer 2: I'm wondering if this should be a function that takes the value and returns a string, since if someone wants custom value text, they probably also want it for the min/max values.

If that's already your plan, then ignore this 😄

Author: What I was thinking for the `valueLabel` prop is that it would handle the two previous API features: `valueFormat` and `showValue`.
If someone wanted to render a valueLabel they could:
1. Pass the Boolean `true` and receive the currentValue.
2. Pass a method like this to customize it:
```jsx
const sliderValue = (value: number) => `${value}%`;
```

Reviewer 2: Got it, thanks for the explanation! I still wonder if it might be better to split those out though. For example, what if someone wanted to customize the min/max labels + maybe the tick labels, but not show separate text for the current value?

| vertical          | &check;                                                                                 | &check;                                                                               | Whether to render the **Slider** vertically. @default `false` (render horizontally)                           |
| marks             | x                                                                                       | x                                                                                     | Whether the **Slider** will have marks to visibly display its steps. @default `false` (renders without marks) |
| fluid             | x                                                                                       | &check;                                                                               | A **Slider** can take the width of its container. @default `false` (width does not fill the container)        |

> | disabledFocusable | X                                                                                       | X                                                                                     | A **Slider** can be disabled and focusable at the same time.                                                  |

Reviewer 2: This seems weird to me -- `disabledFocusable` makes sense for a button or menuitem, but why is this an option for a Slider?

Author: The accessibility design mentioned that the `disabled` Slider should `allow` focus but be readonly to assist screen readers in determining if an element was disabled. The `disabledFocusable` event could be added to follow similar patterns in other converged components. Out of curiosity, how would you suggest handling the `disabled` behavior of the Slider? Should it allow focus in the first place?

| -------- | --------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------- |
| onChange | &check;                                                                                 | &check;                                                                               | Triggers a callback when the value has been changed. This will be called on every individual step. |

### Accessibility props

| Name           | <img src="https://img.shields.io/badge/Used%20in-v0-orange" alt="drawing" width="120"/> | <img src="https://img.shields.io/badge/Used%20in-v8-blue" alt="drawing" width="120"/> | Description                                                           |
| -------------- | --------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------- | --------------------------------------------------------------------- |

> | ariaValueLabel | x                                                                                       | &check;                                                                               | The **Slider's** current value label to be read by the screen reader. |

Reviewer 2: I'm not sure we should specifically support overriding the `valueLabel` (i.e. `aria-valuetext`, I assume) for screen readers. I can't think of a good use case, and I see stuff like this misused quite a lot. The information screen reader users get about the min, max, and value should match what's visually on the screen.

- _Interaction_
  - _Keyboard_
    | Key | Description |
    | - | - |
    | `Up`/`Right` | Increments the value of the slider by the `step` prop. |
    | `Down`/`Left` | Decrements the value of the slider by the `step` prop.|
    | `PageUp`/`Up/Right & Shift`| Increments the value of the slider by 10 _ `step`. |
    | `PageDown` /`Down/Left & Shift`| Decrements the value of the slider by 10 _ `step`. |
    | `Home` | Sets value to the `min` prop. |
    | `End` | Sets value to the `max` prop. |
  - _Cursor_
    `mousedown` sets the current value immediately.
    `mousemove` is attached to the window element on `mousedown` and watches for move events. Value is updated accordingly.
    `mouseup` removes the `mousemove` event.
  - _Touch_
    Handles the same events as the _Cursor_
  - _Screen readers_
    - **`root`:**
      - renders `as` div
      - handles native props expected from the element type in `as`
    - **`thumb`:**
      - `tabindex = 0`
      - `role = "slider"`
      - `aria-valuemin = min` and `aria-valuemax = max`
      - `aria-valuenow` represents the current value.

## Accessibility

> `Slider` design pattern: [W3 Slider](https://www.w3.org/TR/wai-aria-1.1/#slider)

Reviewer 2: The APG actually isn't a good reference for how to build a slider. The purpose of the APG is to show how to use ARIA, but not whether it should be used. In cases where an actual HTML element exists (and especially in this one), it's much, much better to use the HTML element.

Here's a (very old, sorry 😅) demo of creating a custom track, thumb, and fill on top of an `<input type="range">`: https://codepen.io/smhigley/pen/ObWbdy

Author: Thanks for sharing the codepen!

# @fluentui/react-slider Draft Component Spec

## Epic Issue

https://github.com/microsoft/fluentui/issues/18886

## Slider Issues

https://github.com/microsoft/fluentui/issues?q=is%3Aissue+is%3Aopen++is%3A+Component%3A+Slider+

## Background

The Slider component allows users to quickly select a value (or range) by dragging an icon across a bar. It is often used when setting values with a relaxed precision such as audio volume and screen brightness.

## Prior Art

### Open UI Research

https://open-ui.org/components/slider.research.parts
https://open-ui.org/components/slider.research

#### Research Summary

**Marks**: Amongst other component libraries marks/ticks/notches are used to help visibly differ the current location of the thumb. Marks are also used to create custom steps through providing an array of values to jump too.

## Sample Code

```jsx=
// Slider can be uncontrolled
<Slider defaultValue={3} />

// Slider can be controlled
<Slider value={ 3 } />

// The value can be an array for range slider scenarios
<Slider value={ [3, 2] } />

> <Slider defaultValue={ [3, 2] } />

Reviewer: If we are proposing a separate `RangedSlider` we should indicate so in the spec and move this out unto the `RangedSlider` spec.

> <Slider value={ 3 } />
> 
> // The value can be an array for range slider scenarios
> <Slider value={ [3, 2] } />
> <Slider defaultValue={ [3, 2] } />
> 
> // Marks can be a boolean (default marks)
> <Slider marks />
> 
> // Marks can be a number array (specific marks)
> <Slider marks={ [ 2, 5 ] } />

Reviewer: nit:

```suggestion
<Slider value={3} />

// The value can be an array for range slider scenarios
<Slider value={[3, 2]} />
<Slider defaultValue={[3, 2]} />

// Marks can be a boolean (default marks)
<Slider marks />

// Marks can be a number array (specific marks)
<Slider marks={[2, 5]} />
```

// Marks can be an array of mark definitions
<Slider
  marks={ [
    { value: 2, label: "hello" },
    { value: 5, label: "test" }
  ] }
/>

// Marks can be used as the step values using `step`
<Slider step="marks" marks={[25, 50, 75]} />
```

## Variants

- `vertical` displayed vertically.
- `fluid` takes up the entire width of the container.

> - `disabled` read-only and doesn't update state.

Reviewer: nit:

```suggestion
- `disabled` read-only and does not update state.
```

      <slots.rail {...slotProps.rail} />
        {state.marks && (
            <slots.markContainer {...slotProps.markContainer}>
              <slots.mark {...slotProps.mark} />
              {state.markLabel && <slots.markLabel {...slotProps.markLabel} />}
           </slots.markContainer>
      <slots.track {...slotProps.track} />
      <slots.thumb {...slotProps.thumb} />
    </slots.sliderContainer>
    {state.valueLabel && <slots.valueLabel {...slotProps.valueLabel} />}
  </slots.root>
  ```

- _**DOM** - how the component will be rendered as HTML elements_

```jsx
    <div className="ms-Slider-root">
        // Label is up for discussion
        <div className="ms-Slider-label">
            {...label children}
        </div>
        <div className="ms-Slider-container">
            <div className="ms-Slider-rail" />
            <div className="ms-Slider-markContainer">
                <div className="ms-Slider-mark" />
                <div className="ms-Slider-markLabel" />
            </div>
            <div className="ms-Slider-track" />
            <div className="ms-Slider-thumb" />
        </div>
        <div className="ms-Slider-valueLabel" />
    </div>

> ```

Reviewer: I would specially like to see what @smhigley thinks of the structure proposed here since I know that we might not want everything to be a div due to accessibility purposes (maybe we need to render a non-visible input).

>         <div className = "ms-Slider-label">
>             {...label children}
>         </div>
>         <div className="ms-Slider-container">
>             <div className = "ms-Slider-rail" />
>             <div className = "ms-Slider-track" />
>             <div className = "ms-Slider-thumb" />
>             <div className="ms-Slider-markContainer">
>                 <div className="ms-Slider-mark"/>
>                 <div className="ms-Slider-markLabel"/>
>             <div />
>         </div>
>         <div className = "ms-Slider-valueLabel"/>

Reviewer 3: ```suggestion
        <div className="ms-Slider-label">
            {...label children}
        </div>
        <div className="ms-Slider-container">
            <div className="ms-Slider-rail" />
            <div className="ms-Slider-track" />
            <div className="ms-Slider-thumb" />
            <div className="ms-Slider-markContainer">
                <div className="ms-Slider-mark" />
                <div className="ms-Slider-markLabel" />
            </div>
        </div>
        <div className="ms-Slider-valueLabel" />
```

(nit)

    | `PageUp`/`Up/Right & Shift`| Increments the value of the slider by 10 _ `step`. |
    | `PageDown` /`Down/Left & Shift`| Decrements the value of the slider by 10 _ `step`. |
    | `Home` | Sets value to the `min` prop. |
    | `End` | Sets value to the `max` prop. |
  - _Cursor_
    `mousedown` sets the current value immediately.
    `mousemove` is attached to the window element on `mousedown` and watches for move events. Value is updated accordingly.
    `mouseup` removes the `mousemove` event.
  - _Touch_
    Handles the same events as the _Cursor_
  - _Screen readers_
    - **`root`:**
      - renders `as` div
      - handles native props expected from the element type in `as`
    - **`thumb`:**
      - `tabindex = 0`
      - `role = "slider"`
      - `aria-valuemin = min` and `aria-valuemax = max`
      - `aria-valuenow` represents the current value.

## Accessibility

`Slider` design pattern: [W3 Slider](https://www.w3.org/TR/wai-aria-1.1/#slider)

## Topics for Discussion

### Removing Fluid

Having the `Slider` fit the width of the screen could possibly be done by the user by applying custom styling to the `Slider`.

> ### RangedSlider Seperated

Reviewer: nit:

```suggestion
### RangedSlider Separated
```

> It may be more readable and potentially better to split the `RangedSlider` and `Slider` components apart from each other as they have very different use cases.

Reviewer: nit:

```suggestion
It may be more readable and potentially better to split the `RangedSlider` and `Slider` components apart from each other as they have very different use cases and accessibility concerns.
```

### Removing Label and ValueLabel

> If there is a way to remove the Label and ValueLabel while preserving accessibility it could greatly benefit the render structure (Removes container, label, and valueLabel slots).

Reviewer: Again, especially interested of what @smhigley thinks of here.

Reviewer 2: I'm not sure what the purpose of `valueLabel` is, specific to accessibility. I thought it was a visual thing, e.g. if someone wanted text showing the current value to be present, and update as you changed the value?

Label IMO should stay internal, but that seems like a separate discussion.
